### ISSUE TITLE
Start a new instance without a pre-initialized datadir test failure

### ISSUE THREAD

**Issue by cbaines** [eradman/ephemeralpg#1]
Hey, first thanks for writing this, it looks pretty useful!

I've had a quick go at running the included tests, but one of them seems to fail.

```
→ ruby test.rb
0.001: Fetch a random, unused port
0.001: Ensure a new port is picked each time
0.004: Catch unknown options
0.004: Bogus arguments mixed with valid positional
0.004: Run with missing Postgres binaries
0.018: Create a new database on disk
0.028: Start a new instance with a specified datadir and multiple options
0.028: Start a new instance on a TCP port using a specified datadir
Traceback (most recent call last):
	3: from test.rb:133:in `<main>'
	2: from test.rb:12:in `try'
	1: from test.rb:151:in `block in <main>'
test.rb:22:in `eq': "Start a new instance without a pre-initialized datadir" (RuntimeError)
> nice ./pg_tmp initdb
> nice ./pg_tmp -w 60 -d /tmp/ephemeralpg-test.gc7Xw8/ephemeralpg.012345 -p 5432 stop
< nice ./pg_tmp -w 60 -d /tmp/ephemeralpg-test.gc7Xw8/ephemeralpg.012345 -p 5432 stop
< nice ./pg_tmp initdb
```

I'm guessing that the operations are happening in the reverse order from the expectation in the test. Any ideas?

Thanks,

Chris

**Comment by eradman** [comment]
I don't expect the tests to fail.

What platform are you running on?
What is the output of `which psql`?

**Comment by cbaines** [comment]
> What platform are you running on?

I've been trying this out on GNU Guix, a GNU/Linux distro.

> What is the output of `which psql`?

```
→ which psql
/gnu/store/zbm8ilx0yh4a5bhzga4qf1n1cr65c5a4-profile/bin/psql
```

So it's on the `$PATH`, and running `pg_tmp` and `pg_tmp initdb` appear to me to work as well in the same environment within which I'm running the tests (with `make test`).

**Comment by eradman** [comment]
I'll see if I can reproduce this

I downloaded the guix VM image. How do I start `sshd`?

https://guix.gnu.org/manual/en/html_node/Running-Guix-in-a-VM.html#Running-Guix-in-a-VM

I ran `guix install openssh`

**Comment by cbaines** [comment]
> I downloaded the guix VM image. How do I start `sshd`?

That's quite committed! There is a small amount of information in the docs about this [1], and I think what you'll need to do is edit `/etc/config.scm`, add `(service openssh-service-type)` to the list of services, and then run `guix system reconfigure /etc/config.scm` to apply the change. `herd status ssh-daemon` as root will tell you the status of the service.

1: https://guix.gnu.org/manual/en/html_node/Running-Guix-in-a-VM.html#Connecting-Through-SSH

Do let me know if you get stuck, I'm quite happy to dig in more if you think this is something related to Guix.

**Comment by eradman** [comment]
Thanks!  Those steps to configuring system services are what I needed.

I didn't get as far as you did when trying to run the tests

```
Traceback (most recent call last):
        3: from ./test.rb:72:in `<main>'
        2: from ./test.rb:12:in `try'
        1: from ./test.rb:75:in `block in <main>'
./test.rb:22:in `eq': "Run with missing Postgres binaries" (RuntimeError)
> ./pg_tmp: line 29: getopt: command not found
...
```

This is because this test sets `PATH` to `/bin:/usr/bin`. This is a good assumption on platforms with a traditional file system layout, not so good here.  Perhaps I need to discover this path using

```
dirname $(which getopt)
```

**Comment by eradman** [comment]
As for the issue you first reported, this is resolved by changing the behavior of sort(1)

```
LANG=C make test
```

This failure could happen on any distribution; I'll push a fix for this

**Comment by eradman** [comment]
Fixed these issues in commit 67228f3

**Comment by cbaines** [comment]
> Fixed these issues in commit [67228f3](https://github.com/eradman/ephemeralpg/commit/67228f3fb3c3b907cd7db6c305b2f756911b79e8)

Wonderful, thanks so much @eradman !
